Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp

```diff
@@ -383,81 +392,140 @@ namespace Multiplayer
             auto hintSize = aznumeric_cast<uint32_t>(preAliasedSpawnableData.assetHint.size());
 
-            byteStream.Write(sizeof(AZ::Data::AssetId), reinterpret_cast<const void*>(&preAliasedSpawnableData.assetId));
-            byteStream.Write(sizeof(uint32_t), reinterpret_cast<void*>(&hintSize));
-            byteStream.Write(preAliasedSpawnableData.assetHint.size(), preAliasedSpawnableData.assetHint.data());
-            AZ::Utils::SaveObjectToStream(byteStream, AZ::DataStream::ST_BINARY, preAliasedSpawnableData.spawnable.get(), preAliasedSpawnableData.spawnable->GetType());
+            m_levelSendData.m_byteStream->Write(sizeof(AZ::Data::AssetId), reinterpret_cast<const void*>(&preAliasedSpawnableData.assetId));
+            m_levelSendData.m_byteStream->Write(sizeof(uint32_t), reinterpret_cast<void*>(&hintSize));
+            m_levelSendData.m_byteStream->Write(preAliasedSpawnableData.assetHint.size(), preAliasedSpawnableData.assetHint.data());
+            AZ::Utils::SaveObjectToStream(
+                *m_levelSendData.m_byteStream,
+                AZ::DataStream::ST_BINARY,
+                preAliasedSpawnableData.spawnable.get(),
+                preAliasedSpawnableData.spawnable->GetType());
         }
         
         // Spawnable library needs to be rebuilt since now we have newly registered in-memory spawnable assets
         AZ::Interface<INetworkSpawnableLibrary>::Get()->BuildSpawnablesList();
 
         // Read the buffer into EditorServerLevelData packets until we've flushed the whole thing
-        byteStream.Seek(0, AZ::IO::GenericStream::SeekMode::ST_SEEK_BEGIN);
+        m_levelSendData.m_byteStream->Seek(0, AZ::IO::GenericStream::SeekMode::ST_SEEK_BEGIN);
 
         // Send an initial notification showing how much data will be sent.
         MultiplayerEditorServerNotificationBus::Broadcast(
             &MultiplayerEditorServerNotificationBus::Events::OnEditorSendingLevelData,
-            0, aznumeric_cast<uint32_t>(byteStream.GetLength()));
+            0,
+            aznumeric_cast<uint32_t>(m_levelSendData.m_byteStream->GetLength()));
+
+        // The actual data will get sent "asynchronously" during the OnTick callback over multiple frames.
+    }
+
+    void MultiplayerEditorSystemComponent::SendLevelDataToServer()
+    {
+        // This controls the maximum time slice to use for sending packets. Lower numbers will make the total send time take longer,
+        // but will give the Editor more time to do other work. Larger numbers will make the total send time faster, but will starve
+        // the Editor. The current value attempts to balance between the two.
+        static constexpr AZ::TimeMs MaxSendTimeMs = AZ::TimeMs{ 5 };
+
+        // These control how many retries and how to space them out for packet send failures.
+        static constexpr int MaxRetries = 20;
+        static constexpr int InitialMsDelayPerRetry = 10;
+        static constexpr int MaxMsDelayPerRetry = 1000;
```

Comment:
Shouldn't these variables be of type AZ::TimeMs?

Reply:
since they're getting passed into chrono functions that take in ints, it seemed like int was the right type?

Reply:
Well they get passed to an `AZStd::chrono::milliseconds` constructor which has a representation type for the integer type that is at least [45 bits](https://en.cppreference.com/w/cpp/chrono/duration#Helper_types).

It shouldn't be a problem though since the millisecond constants here are able to easily fit into an 32-bit int